zkevm-circuits/src/evm_circuit/execution/signed_comparator.rs

```diff
@@ -0,0 +1,301 @@
+        // result.
+        // result = if a < 0 && b >= 0, slt = 1.
+        // result = if b < 0 && a >= 0, slt = 0.
+        let a_neg_b_pos = (1.expr() - a_pos.expr()) * b_pos.expr();
+        let b_neg_a_pos = (1.expr() - b_pos.expr()) * a_pos.expr();
+        let result = select::expr(
+            a_neg_b_pos,
+            1.expr(),
+            select::expr(b_neg_a_pos, 0.expr(), a_lt_b),
+        );
+
+        // Pop a and b from the stack, push the result on the stack.
+        cb.stack_pop(select::expr(is_sgt.expr(), b.expr(), a.expr()));
+        cb.stack_pop(select::expr(is_sgt.expr(), a.expr(), b.expr()));
+        cb.stack_push(result);
+
+        // The read-write counter changes by three since we're reading two words
+        // from stack and writing one. The program counter shifts only by one
+        // and the since the stack now has one less word, the stack pointer also
+        // shifts by one.
+        let step_state_transition = StepStateTransition {
+            rw_counter: Transition::Delta(3.expr()),
+            program_counter: Transition::Delta(1.expr()),
+            stack_pointer: Transition::Delta(1.expr()),
+            ..Default::default()
+        };
+        let same_context = SameContextGadget::construct(
+            cb,
+            opcode,
+            step_state_transition,
+            None,
+        );
+
+        Self {
+            same_context,
+            a,
+            b,
+            sign_check_a,
+            sign_check_b,
+            lt_lo,
+            comparison_hi,
+            is_sgt,
+        }
+    }
+
+    fn assign_exec_step(
+        &self,
+        region: &mut Region<'_, F>,
+        offset: usize,
+        block: &Block<F>,
+        _transaction: &Transaction<F>,
+        _call: &Call<F>,
+        step: &ExecStep,
+    ) -> Result<(), Error> {
+        self.same_context.assign_exec_step(region, offset, step)?;
+
+        let opcode = step.opcode.unwrap();
+
+        // SLT opcode is the default check in the SCMP gadget. Swap rw for SGT.
+        self.is_sgt.assign(
+            region,
+            offset,
+            F::from(opcode.as_u8() as u64),
+            F::from(OpcodeId::SGT.as_u8() as u64),
+        )?;
+        let indices = if opcode == OpcodeId::SGT {
+            [step.rw_indices[1], step.rw_indices[0]]
+        } else {
+            [step.rw_indices[0], step.rw_indices[1]]
+        };
+        let [a, b] =
+            indices.map(|idx| block.rws[idx].stack_value().to_le_bytes());
+
+        // Assign to the sign check gadgets. Since both a and b are in the
+        // little-endian form, the most significant byte is the last byte.
+        self.sign_check_a.assign(
+            region,
+            offset,
+            F::from(a[31] as u64),
+            F::from(128u64),
+        )?;
+        self.sign_check_b.assign(
+            region,
+            offset,
+            F::from(b[31] as u64),
+            F::from(128u64),
```

Comment:
Sorry to ask. But unsure why `rhs` is passed as 128.

Reply:
@CPerezz The values `a` and `b` are in the two's complement form, representing signed integers. So for the most significant byte (MSB), i.e. `a[31]` or `b[31]`, the first bit is set to `1` to indicate that the value `a` or `b` is negative.

If that bit is `1`, it means the `MSB >= 128`. So by checking if `MSB < 128`, we know that the value of `a` or `b` itself is `>= 0`.

Reply:
That makes complete sense! 
Apologies for the question :sweat_smile: